### ISSUE TITLE
Week 1

### ISSUE THREAD

**Issue by Seungjun1127** [Seungjun1127/Mamba-ITS#1]
**Sprint 1 Goal:** Establish the development environment, thoroughly understand the integration points of the Swin Transformer within ViTST, understand the Vim model's interface, and create a basic code structure replacing Swin with Vim (even if non-functional initially).

**Key Repositories/Codebases:**
1.  **ViTST:** `https://github.com/Leezekun/ViTST` (from the paper)
2.  **Vim:** `https://github.com/hustvl/Vim` (from the paper)

---

**Week 1 Plan (Sprint 1)**

**Day 1: Setup & Initial Exploration**

*   **Morning (Both Devs - Pair/Collaborate):**
    *   **Task 1.1:** Agree on and set up the shared Python environment (e.g., using `conda`).
        *   `conda create -n vitst_vim python=3.9` (or appropriate version)
        *   `conda activate vitst_vim`
    *   **Task 1.2:** Install essential base libraries: `pytorch`, `torchvision`, `torchaudio` (matching CUDA version if applicable), `matplotlib`, `numpy`, `scikit-learn`.
    *   **Task 1.3:** Clone the ViTST repository: `git clone https://github.com/Leezekun/ViTST.git`
    *   **Task 1.4:** Clone the Vim repository: `git clone https://github.com/hustvl/Vim.git` 
    *   **Task 1.5:** Install ViTST's specific dependencies (check its `requirements.txt` or setup instructions).
    *   **Task 1.6:** Install Vim's specific dependencies (check its requirements). May involve custom CUDA kernels if using Mamba components directly – flag potential build issues early.
    *   **Goal:** Both developers have identical, functional environments and have cloned both codebases.
*   **Afternoon (Split Tasks):**
    *   **Task 2.1 (Dev A):** Begin ViTST code exploration.
        *   Run the existing ViTST code on a small dummy dataset or a few samples from PAMAP2 (if easily runnable) to see it working with Swin.
        *   Identify the main model definition file (likely involves searching for `SwinTransformer` import/usage).
        *   Start tracing the `forward` pass: How does data get from the input image grid to the Swin backbone?
    *   **Task 3.1 (Dev B):** Begin Vim code exploration.
        *   Understand the Vim repository structure. Find the main model definition (`vim.py`?).
        *   Look for example usage scripts (e.g., for ImageNet classification). Try running a basic example if available.
        *   Identify how the `Vim` model class is instantiated. What are the key parameters (`img_size`, `patch_size`, `embed_dim`, `depths`, etc.)?

**Day 2: Deep Dive Analysis**

*   **Morning (Split Tasks):**
    *   **Task 2.2 (Dev A):** Continue ViTST deep dive.
        *   Pinpoint the *exact* lines where `SwinTransformer` is instantiated. Note the parameters being passed (pretrained weights path, image size, patch size, window size, embed dim, depths, num_heads).
        *   Analyze the input tensor passed to Swin's `forward` method: What is its shape? (Likely `[Batch, Channels, Height, Width]` or `[Batch, Num_Patches, Embed_Dim]`).
        *   Analyze the output tensor from Swin: What is its shape? How is it processed before the final classification layer (e.g., pooling, LayerNorm, specific token extraction)?
        *   Check ViTST configuration files (if any) for backbone settings.
    *   **Task 3.2 (Dev B):** Continue Vim deep dive.
        *   Determine Vim's expected input format precisely. Does it take a `[B, C, H, W]` image and do patching internally, or does it expect `[B, N, D]` sequence of patch embeddings? (Crucial for integration). The `hustvl/Vim` repo likely expects `[B, N, D]`.
        *   Determine Vim's output format. Does it return a sequence `[B, N, D]`? Does it have a specific classification token output?
        *   Figure out how to load pre-trained weights. Is there a helper function? Where are the official checkpoints hosted? Download a relevant checkpoint (e.g., `vim_s_mid_224_1k.pth`).
*   **Afternoon (Collaborate & Document):**
    *   **Task 5.1 (Both Devs):** Sync up. Discuss findings about Swin's interface within ViTST and Vim's interface requirements. Identify key points of mismatch (e.g., input/output shapes, patching strategy).
    *   **Task 5.2 (Both Devs):** Start shared documentation (e.g., `integration_notes.md`). Record:
        *   Swin instantiation details in ViTST.
        *   Swin input/output shapes observed.
        *   Vim instantiation parameters.
        *   Vim expected input/output shapes.
        *   How to load Vim pretrained weights (initial understanding).
        *   *Key Questions/Challenges:* How to handle patching? How to adapt output dimensions?

**Day 3: Structural Integration**

*   **Morning (Pair Programming Recommended):**
    *   **Task 4.1:** Create the feature branch: `git checkout -b feat/vim-integration`.
    *   **Task 4.2:** Open the ViTST model definition file.
    *   **Task 4.3:** Comment out the `SwinTransformer` import and instantiation lines.
    *   **Task 4.4:** Add the `Vim` import from the Vim library.
    *   **Task 4.5:** Add code to instantiate the `Vim` model. Use parameters that seem compatible or are placeholders for now (e.g., matching embedding dimension if possible, placeholder image/patch size). *Don't* try to load weights yet.
        ```python
        # Example placeholder
        # from vim.models_vim import Vim # Adjust import path as needed
        # self.vim_backbone = Vim(embed_dim=128, depth=..., ...) # Use some parameters
        ```
    *   **Task 4.6:** Modify the `forward` pass. Comment out the line calling Swin (`output = self.swin(...)`). Add a line calling Vim (`output = self.vim_backbone(...)`).
        *   **Crucial:** Pass the *same input variable* to Vim as was passed to Swin for now. Expect errors later, the goal is just structural replacement.
    *   **Task 4.7:** Ensure the file saves and *imports* without syntax errors. Run `python path/to/model_file.py` if it's executable standalone, or try importing the model class in a separate test script.
*   **Afternoon (Split Tasks):**
    *   **Task 2.3 (Dev A):** Refine understanding of ViTST's patching. How are the grid images turned into patches *before* Swin? Is it a standard ViT-style patch embedding layer? What's the output shape of this patching step?
    *   **Task 3.3 (Dev B):** Write a small, standalone script (`test_vim_load.py`) to:
        *   Instantiate the Vim model (using parameters from its standard config, e.g., for `vim_s_mid_224_1k`).
        *   Load the downloaded pre-trained weights (`vim_s_mid_224_1k.pth`). Debug any state dict key mismatches or loading errors.
        *   Create a dummy input tensor matching Vim's *expected* input shape (e.g., `torch.randn(1, 197, 768)` if it's ViT-like).
        *   Pass the dummy input through the loaded Vim model. Verify it runs and get the output shape.

**Day 4: Refining Stubs & Addressing Mismatches**

*   **Morning (Split Tasks):**
    *   **Task 1.7 (Dev A):** Based on Task 2.3 findings, analyze if ViTST's *existing* patch embedding mechanism can produce the input shape required by Vim (Task 3.2/3.3). If not, what modifications are needed? (e.g., change projection dimension, add/remove class token, flatten/reshape). *Sketch out* the required adapter code conceptually or in comments.
    *   **Task 1.8 (Dev B):** Analyze the output shape from the standalone Vim test (Task 3.3). Compare it to the shape Swin provided *to* ViTST's classification head (Task 2.2). What adapter is needed? (e.g., select a specific token, apply pooling, use a Linear layer to adjust dimensions). *Sketch out* the required adapter code conceptually or in comments.
*   **Afternoon (Pair Programming Recommended):**
    *   **Task 4.8:** In the ViTST model file (`feat/vim-integration` branch), *implement* the input adapter sketched by Dev A. Modify the code *before* the `self.vim_backbone(...)` call to ensure the tensor has the correct shape/format for Vim. Use `print(tensor.shape)` liberally for debugging.
    *   **Task 4.9:** *Implement* the output adapter sketched by Dev B. Modify the code *after* the `self.vim_backbone(...)` call to ensure the tensor has the shape expected by the subsequent layers (likely the classification head). Use `print(tensor.shape)`.
    *   **Goal:** Have placeholder adapters in place. Still expect runtime errors, but the *intended flow* connecting ViTST patcher -> Vim -> ViTST head is now structurally coded.

**Day 5: Basic Run & Wrap-up**

*   **Morning (Pair Programming Recommended):**
    *   **Task 5.3:** Try to run a *single forward pass* with a *single batch* of real data from ViTST's data loader (e.g., from PAMAP2).
        *   Set batch size to 1 or 2 for easier debugging.
        *   Add breakpoints or extensive print statements to trace tensor shapes through: `DataLoader -> Image Generation -> Patching -> Input Adapter -> Vim -> Output Adapter -> Head`.
        *   Identify and fix the first few obvious runtime errors (shape mismatches, dtype mismatches). **Do not** aim for a fully working forward pass yet, just make progress.
*   **Afternoon (Both Devs):**
    *   **Task 5.4:** Clean up code committed to `feat/vim-integration`. Ensure it's readable.
    *   **Task 5.5:** Finalize the `integration_notes.md` documentation with detailed findings, implemented adapter logic (even if partial), and remaining challenges/next steps for Sprint 2.
    *   **Task 5.6:** Prepare for Sprint Review:
        *   Demo the environment setup.
        *   Show the modified ViTST model code with Vim structurally integrated and placeholder adapters.
        *   Show the standalone Vim loading script working.
        *   Discuss the findings regarding input/output mismatches.
        *   Present the documentation/notes.
        *   Briefly outline the plan/challenges for Sprint 2 (making the forward/backward pass fully functional).
    *   **Task 5.7:** Sprint Retrospective (briefly discuss what went well, what didn't, what to improve for Sprint 2).

---

**Definition of Done (DoD) for Sprint 1 (Revisited):**

*   ✅ Shared conda environment is functional and documented.
*   ✅ ViTST and Vim codebases cloned.
*   ✅ ViTST model file modified on `feat/vim-integration` branch: Swin commented out, Vim instantiated, forward pass structurally calls Vim.
*   ✅ Placeholder input/output adapter logic implemented between ViTST components and Vim.
*   ✅ Standalone script successfully loads pre-trained Vim weights and performs inference on dummy data.
*   ✅ Shared documentation details Swin/Vim interfaces, integration points, and initial challenges.
*   ✅ Sprint Review demo prepared, showing progress and findings.
*   ✅ Code imports correctly (runtime errors during full forward pass are expected).

This detailed plan provides concrete daily goals. Flexibility is still needed – if a task takes longer, adjust accordingly, but always keep the main Sprint Goal in sight. Good communication will be key!

**Comment by Tosaaa** [comment]
어 근데 저희 MambaVision 쓰는거 아니었나요?
Midterm proposal에서도 MambaVIsion으로 발표했었어요

**Comment by Seungjun1127** [comment]
그러게요 우선 Notion에 정리된 논문 기준으로 작성했는데 MambaVision으로 가면 될거같아요
